core/domain/interaction_validation_jobs_one_off.py

```diff
@@ -0,0 +1,278 @@
+                                ' Rule: %s is invalid.' % (rule_index) +
+                                '(Indices here are 0-indexed.)')
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, values)
+
+
+class ItemSelectionInteractionOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of (exploration, state) pairs that use the item
+    selection interaction and that have rules that do not match the answer
+    choices. These probably need to be fixed manually.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        exploration = exp_fetchers.get_exploration_from_model(item)
+        for state_name, state in exploration.states.items():
+            if state.interaction.id == 'ItemSelectionInput':
+                choices = (
+                    state.interaction.customization_args['choices']['value'])
+                for group in state.interaction.answer_groups:
+                    for rule_spec in group.rule_specs:
+                        for rule_item in rule_spec.inputs['x']:
+                            if rule_item not in choices:
+                                yield (
+                                    item.id,
+                                    '%s: %s' % (
+                                        state_name.encode('utf-8'),
+                                        rule_item.encode('utf-8')))
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, values)
+
+
+class InteractionRTECustomizationArgsValidationOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for validating all the customizations arguments of
+    Rich Text Components.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+        err_dict = {}
+
+        try:
+            exploration = exp_fetchers.get_exploration_from_model(item)
+        except Exception as e:
+            yield (
+                'Error %s when loading exploration'
+                % python_utils.UNICODE(e), [item.id])
+            return
+
+        html_list = exploration.get_all_html_content_strings()
+        err_dict = html_validation_service.validate_customization_args(
+            html_list)
+
+        for key in err_dict:
+            if err_dict[key]:
+                yield ('%s Exp Id: %s' % (key, item.id), err_dict[key])
+
+    @staticmethod
+    def reduce(key, values):
+        final_values = [ast.literal_eval(value) for value in values]
+        # Combine all values from multiple lists into a single list
+        # for that error type.
+        output_values = list(set().union(*final_values))
+        exp_id_index = key.find('Exp Id:')
```

Comment:
I think you can skip this and just do `if ... in key`?

Reply:
The index is required in else part.

Reply:
Actually, I'm a bit confused by this. It seems to me that this reduce() runs once per key, and the key is specific to a particular type of error and a particular exp ID. So this isn't actually going to combine stuff from multiple exp IDs into one item. Is that intended?

(I don't think your backend tests check the case of multiple exps with the same error -- might want to add one.)

Reply:
I think the purpose is to get all errors of one type (example: errors related to invalid url) as a set irrespective of the exploration since exp id might not be useful in debugging these errors as such. So, I think the approach is intended.
@bansalnitish to confirm since he [added this job](https://github.com/oppia/oppia/commit/f69ec711e055654f1488feca042f652af4aca300#diff-e4833416cf9d665369010673d2fba853R819) in the first place. I am adding the tests as well.

Reply:
I'm fine with that but in that case shouldn't the exp ID be in the 2nd part of the tuple returned in map()?

Reply:
Just checked a bit more, the exp id is being added to the error message - not sure what is the purpose of that. @bansalnitish any thoughts?

Reply:
Hmm so I think it is structured in this way:

Key: Error type along with what is the invalid data
Value: Complete html where invalid data is found, along with Exploration ID

Example:

Key: Invalid url, received "some url"
Value: <oppia-..link value=some url..></oppia..link>, Exp ID: exp_id1

Reply:
That's what I think it should be, too, but the yielded value at the end of map() does not seem to match that.

Reply:
Added test case for multiple explorations.

Reply:
It matches that, right? What part is the mismatch?

Reply:
I'm reading `yield ('%s Exp Id: %s' % (key, item.id), err_dict[key])`

So -- isn't the first item in this tuple `{{error type}} Exp Id: {{exp id}}`, and the second item a list of html instances in the exploration that exhibit this error?

If so, then reduce() would group on the first item and get all the values that map to it, then combine those only. It doesn't know about other tuples with different exp IDs because they would result in different keys. So if the same error type appears in different explorations, the corresponding html strings wouldn't be combined.

Or at least, that's my reasoning. What am I missing?

Reply:
Yes, you are right and I agree we should not have messages per exp rather they should be per error category. I have updated the output generation according to that.